doc: add ability to only generate kernel documentation

dcpleung opened the pull request:
This adds a few targets to generate core kernel documentation
(including the Kernel Primer and API).

Relates to #6782

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dbkinder commented on doc/kernel.rst.kerneldoc:
```diff
@@ -0,0 +1,33 @@
+Application Development Primer
+""""""""""""""""""""""""""""""
+
+https://docs.zephyrproject.org/latest/application/application.html
```
This could use a sentence around the document link rather than just a bare link.

dbkinder commented on .known-issues/doc/kerneldoc/unknown_options.conf:
```diff
@@ -0,0 +1,2 @@
+# since we only parse Kconfig under kernel/, need to ignore those under arch/
```
This "known issues" filter will apply to all documentation builds and could let some errors slip by for normal doc builds.  Perhaps the errors should be removed by adding a few more Kconfig files as part of what we'd consider the "kernel" configurations.

Sure seems like features the kernel relies on such as FP and MPU configuration, could be considered part of the kernel docs.

dcpleung replied:
Both ```arch/Kconfig``` and ```kernel/Kconfig``` are included by ```Kconfig.zephyr``` where they are at the same level in the Kconfig hierarchical "tree". I can't think of a scalable solution without introducing another top level Kconfig file just for document generation. A hierarchy that works for kernel doc may not work for api doc. But adding more files means that someone needs to maintain them.

Another solution would be to parse all the relevant RST files to extra all the referenced Kconfig options. The list of options will be fed into the genrest.py script to generate the kconfig RST files. However, this is error prone as we have to parse RST files ourselves.

Of course, we can also not customize the Kconfig doc and includes everything. HTML would probably be fine (except with more files and linkes), but number of pages for PDF will increase dramatically.

dcpleung replied:
The latest version has a way to specify what the root Kconfig to include necessary options. So this is no longer an issue.

dbkinder commented on doc/conf.py:
```diff
@@ -63,7 +63,10 @@
 # The master toctree document.
-master_doc = 'index'
+if tags.has("kerneldoc"):
+    master_doc = 'kernel'
```
This won't result in an ``index.html`` for the doc root in the generated website.  I'm not sure where you intend to publish this html tree, or is the intention just to create a PDF file for "Kernel Only" documentation?

dcpleung replied:
This will probably be an easy fix to generate the index.html file.

dbkinder commented on doc/kernel.rst.kerneldoc:
```diff
@@ -0,0 +1,33 @@
+   reference/kconfig/index.rst
+
+Links
+^^^^^
```
This isn't following the Zephyr documentation convention for heading hierarchy:  ``# * = -``

dbkinder reviewed (changes requested):
This solution isn't scalable for creating new "slices" of the Zephyr documentation.  Is a "Kernel only" doc set the only foreseeable customized doc set needed?  Not included in this PR (and referenced issue #6782) is the public motivation and need for this change.

dcpleung commented:
Most of the burden is on limiting the content included in the final document, e.g., limiting the Kconfig options included.

Also, some extra bits are needed to avoid broken links. For example, if only the ```kernel/kernel.rst``` is used as the ```master_doc```, references to ```api``` and ```application``` are broken since their respective RST  files are not included. We will need to define these labels somewhere so these references can be used to point to external URLs. However, Sphinx does not make this easy since ```only``` and ```ifconfig``` do not work correctly with labels. So when generating the whole document, there would be two ```api``` labels and Sphinx will complain. This is why ```kernel.rst.kerneldoc``` is introduced. In other words, extracting a sub-tree out of the RST source tree not an easy task without introducing more bits that needs to be maintained.

nashif commented:
I wonder if we can do all of this in .rst directly using conditionals instead of doing this on the build system level. For example we should be able to only include certain chapters

dcpleung commented:
Started from scratch in a less hack-ish way. It should be a lot easier to extend this to other subset of documentation.

@nashif Sphinx does not make it easy to do conditionally includes. From what I have observed, directives like include and toctree are processed anyway. And the ::only or ::ifconfig directives act as if they are post-processing the content. Which means everything in the supposedly excluded files are already parsed and references resolved. With the internal cross-references like :ref:application or :ref:api, you will need to include them as output (e.g. put on the webserver) or else they become invalid links.

carlescufi commented:
@dcpleung we have a change that is about to be merged that will impact your PR here: https://github.com/zephyrproject-rtos/zephyr/pull/10470

I suggest we clarify the merge order, although #10470 is ready to be merged now.

dcpleung commented:
@carlescufi I took a quick look at #10470. It shouldn't be too hard to re-work this PR on top of that. Since that one is ready, merge it and I will rework this PR.

carlescufi reviewed (changes requested):
https://github.com/zephyrproject-rtos/zephyr/pull/10252/files#r224739993

dcpleung commented on doc/CMakeLists.txt:
```diff
@@ -63,166 +56,275 @@ set(I18NSPHINXOPTS  ${SPHINXOPTS})
-endif()
+function(construct_html_target scope)
+  if("${scope}" STREQUAL "")
+    set(SPHINX_OUTPUT_DIR_HTML ${SPHINX_OUTPUT_DIR}/html)
```
Fixed in latest revision.

dcpleung commented:
Rebased onto top of master which included #10470.

carlescufi commented on doc/CMakeLists.txt:
```diff
@@ -65,248 +60,414 @@ set(I18NSPHINXOPTS  ${SPHINXOPTS})
+#
+# Macro to generate target for document content
+#
+# Note: this is a macro because variables set by this
```
You can override that in a `function` by setting the `PARENT_SCOPE` as well. No preference but just to consider.

dcpleung replied:
I am thinking that the extraction scripting will probably get more enhancements in the future. With this being a macro, there is no need to track down variable scope below. *Might* make it easier to maintain in the future.

carlescufi reviewed (commented):
Removing my -1 since I have no additional real objections to this PR. Will leave it to @dbkinder

dcpleung commented on doc/CMakeLists.txt:
```diff
@@ -65,248 +60,414 @@ set(I18NSPHINXOPTS  ${SPHINXOPTS})
+  if("${scope}" STREQUAL "")
+    set(CONTENT_OUTPUTS ${CMAKE_CURRENT_BINARY_DIR}/extracted-content.txt)
+  else()
+    set(CONTENT_OUTPUTS ${CMAKE_CURRENT_BINARY_DIR}/${scope}-extracted-content.txt)
```
The patch has the ${scope} as prefix to output files/directories. I can change it. I just thought that having scope as prefix groups all relevant bits together (ls and tab-completion).

dcpleung replied:
The content extraction has been reverted so this is no longer an issue.

mbolivar commented:
Seems like this needs a rebase onto the final version of https://github.com/zephyrproject-rtos/zephyr/pull/10470

mbolivar commented:
> I wonder if we can do all of this in .rst directly using conditionals instead of doing this on the build system level. For example we should be able to only include certain chapters

Is using intersphinx to separate the kernel into its own docs toctree an option? The rest of the docs could potentially sit on "top" of that and refer into it. Not sure if I'm talking crazy, I've never used intersphinx, just read about its features.

dcpleung commented:
Skimming through intersphinx's page suggests that it requires internet connection to resolve missing references. No?

dcpleung commented:
Rebased.

carlescufi commented:
@dcpleung I have to apologize to you for not having merged this earlier. Turns out we just simplified the doc build system and you'll need to rebase again. Sorry about that!

dcpleung commented:
@carlescufi Sorry for late reply. I was on vacation. :)

I have updated the patch set on top of master with the latest change under doc/.

carlescufi commented on doc/CMakeLists.txt:
```diff
@@ -65,217 +60,387 @@ set(I18NSPHINXOPTS  ${SPHINXOPTS})
+#       are being used in constructing other targets.
+#
+macro(construct_content_target scope)
+  if("${scope}" STREQUAL "")
```
Please remove everything with `CONTENT_OUTPUTS` and the targets that depend on it, such as `sphinx-html` and `sphinx-latex`. Those are no longer valid since we simplified the build system, and are a leftover that should've been removed in the first place.

carlescufi replied:
This is because we no longer generate the list of `.rst` files since that was simply not effective

carlescufi commented:
@dcpleung thanks for your patience. There was some stuff left over after the cleanup that needs to be removed, since it's no longer populated (everything `CONTENT_OUTPUTS` related)

carlescufi commented:
@mbolivar 
>Is using intersphinx to separate the kernel into its own docs toctree an option? The rest of the docs could potentially sit on "top" of that and refer into it. Not sure if I'm talking crazy, I've never used intersphinx, just read about its features.

intersphinx requires 2 different sphinx builds and then connecting them together. That would mean for standard "full" Zephyr doc builds we would need to run sphinx twice, making it even slower than it is today. I think this solution is more practical.

dcpleung committed:
doc: add ability to only generate kernel documentation

This adds a few targets to generate core kernel documentation
(including the Kernel Primer and API).

Relates to #6782

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
doc: Correct LATEXOPTS quotation for Win32 and other systems

On Windows systems, LATEXOPTS needs to be declared with double quotes
for proper operation. However, on other systems, the double quotes
make it a whole string being passed as one command argument instead of
multiple ones. So use single quotes for non-Win32 systems.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
doc: add zephyr.doc_scope extension

This extension enables:

() Filter the document list discovered by Sphinx to limit
   which files are included.
() Goes through the toctree and removes the entries that
   points to document files no longer included.
() Replace the references to removed files with external
   URL.

This is in preparation for enabling only building a subset
of documentation.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
doc: make build target generations into functions

This allows the document build targets to be generated
instead of hard-coding everything. This is in preparation
to enable generating a subset of document (e.g. kernel
only, or api only), where there will be multiple targets
having more or less very similar target definition.

Note that the output locations are changed if
SPHINX_OUTPUT_DIR is externally defined. Now HTML files
will be put under html/, LaTex under latex/, and
PDF under pdf/.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

mbolivar commented on doc/conf.py:
```diff
@@ -259,7 +260,10 @@
 #html_search_scorer = 'scorer.js'
 
 # Output file base name for HTML help builder.
-htmlhelp_basename = 'zephyrdoc'
```
we don't use the HTML help or manpage builders at all as far as I know; why not just remove them?

dcpleung replied:
There definitely are no build targets for these. But since these options are not commented out and have "zephyr" in them, it would be better to also modify them to be consistent with the logic/thinking of this PR (which is simply to add the ability to generate kernel documentation only).

mbolivar commented on doc/conf.py:
```diff
@@ -323,11 +339,18 @@
-   'Miscellaneous'),
-]
+if tags.has("kernel"):
+    texinfo_documents = [
```
Same here; we don't use the texinfo builder, do we?

dcpleung replied:
(See above)

We can probably comment these out in another PR.

mbolivar reviewed (commented):
The approach seems very complex but I'll trust you and @carlescufi that it's the best way. I just have a couple of questions.

dcpleung commented on doc/extensions/zephyr/doc_scope.py:
```diff
@@ -0,0 +1,257 @@
+
+    return {
+        'version': '1.0',
+        'parallel_read_safe': True,
```
According to Sphinx document, during the stage this plugin runs, all the files have been already been read and parsed (but before being pickled). And this plugin only manipulates in-memory structures for each document. So this should be safe.

dcpleung commented:
Rebased due to new changes related to shuffling of files under doc/.

carlescufi commented:
>The approach seems very complex but I'll trust you and @carlescufi that it's the best way. I just have a couple of questions.

@dcpleung @mbolivar 
I am thinking about this again and I still support @dcpleung's approach over intersphinx. The reason is that with the current approach we have complete freedom on how to layout the documentation in a single tree, whereas with intersphinx we'd be forced to split it in separate documents that link together with specific named tags. Also even if we went for the intersphinx way we'd need to deal with a good chunk of the added complexity here (separate Kconfig and content targets), and the default complete build would be further slowed down because of running all targets twice.

@dbkinder I think this is as scalable as it can be, there's a set of scope documents and the framework to add new scopes in case we want to add new targets.

The only way I can see of doing this efficiently and in a simpler way is to overhaul the doc completely so that it is designed from scratch to be built in slices, but even then I am not sure we'd be able to do away with some of the complexity.

galak commented:
@dcpleung is this still relevant?

dcpleung commented:
@galak Depends on if generating only a section of documentation is still desired. Maybe we can close this, and if someone really wants this feature, we can re-open.

galak commented:
> @galak Depends on if generating only a section of documentation is still desired. Maybe we can close this, and if someone really wants this feature, we can re-open.

I say we close it and re-open if needed.

dcpleung commented:
Closed for now. Re-open if needed.
